src/modules/ekf2/EKF/aid_sources/range_finder/range_height_control.cpp

```diff
@@ -270,13 +271,27 @@ float Ekf::getRngVar() const
 
 void Ekf::resetTerrainToRng(estimator_aid_source1d_s &aid_src)
 {
-	const float new_terrain = _state.pos(2) + aid_src.observation;
-	const float delta_terrain = new_terrain - _state.terrain;
+	// Since the distance is not a direct observation of the terrain state but is based
+	// on the height state, a reset should consider the height uncertainty. This can be
+	// done by manipulating the Kalman gain to inject all the innovation in the terrain state
+	// and create the correct correlation with the terrain state with a covariance update.
+	P.uncorrelateCovarianceSetVariance<State::terrain.dof>(State::terrain.idx, 0.f);
 
-	_state.terrain = new_terrain;
-	P.uncorrelateCovarianceSetVariance<State::terrain.dof>(State::terrain.idx, aid_src.observation_variance);
+	const float old_terrain = _state.terrain;
+	_state.terrain = 0.f;
```

Comment:
Questions/Ideas
 - could we/should we skip the reset if the test ratio was fine in the first place?
 - why reset the state to 0? What about resetting to a filtered value (perhaps a median filter for RNG data) and then measurementUpdate the current sample on top?

Reply:
> could we/should we skip the reset if the test ratio was fine in the first place?

That's already the case, we only reset if the test ratio is failing: https://github.com/PX4/PX4-Autopilot/blob/21a2350ecab485726c63d4269986808fc291fa61/src/modules/ekf2/EKF/aid_sources/range_finder/range_height_control.cpp#L218-L220

> why reset the state to 0? What about resetting to a filtered value (perhaps a median filter for RNG data) and then measurementUpdate the current sample on top?

That reset value actually doesn't matter since we push the complete innovation in the terrain state (`K(terrain) = 1`). We could however use a filtered sample to compute the innovation, yes, that might be a good idea.
We don't need to reset and then fuse as this way of doing the fusion already does the reset (without introducing incorrect uncertainties).